buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```
@@ -0,0 +1,1428 @@
+        }
+        // The magazines failed us, or the buffer is too big to be pooled.
+        return chunkAllocator.allocate(size, maxCapacity);
+    }
+
+    public long usedMemory() {
+        long sum = 0;
+        for (ByteBuf byteBuf : centralQueue) {
+            sum += byteBuf.capacity();
+        }
+        for (Magazine magazine : magazines) {
+            sum += magazine.usedMemory.get();
+        }
+        return sum;
+    }
+
+    private boolean tryExpandMagazines(int currentLength) {
+        if (currentLength >= MAX_STRIPES) {
+            return true;
+        }
+        long writeLock = magazineExpandLock.tryWriteLock();
+        if (writeLock != 0) {
+            try {
+                Magazine[] mags = magazines;
+                if (mags.length >= MAX_STRIPES || mags.length > currentLength) {
+                    return true;
+                }
+                Magazine[] expanded = Arrays.copyOf(mags, mags.length * 2);
+                for (int i = mags.length, m = expanded.length; i < m; i++) {
+                    expanded[i] = new Magazine(this);
+                }
+                magazines = expanded;
+            } finally {
+                magazineExpandLock.unlockWrite(writeLock);
+            }
+        }
+        return true;
+    }
+
+    public void close() {
+        closed = true;
+        long magsExpandWriteLock = magazineExpandLock.writeLock();
+        try {
+            for (Magazine mag : magazines) {
+                long writeLock = mag.writeLock();
+                try {
+                    mag.close();
+                } finally {
+                    mag.unlockWrite(writeLock);
+                }
+            }
+        } finally {
+            magazineExpandLock.unlockWrite(magsExpandWriteLock);
+        }
+        drainCloseCentralQueue();
+    }
+
+    private void drainCloseCentralQueue() {
+        ByteBuf curr;
+        while ((curr = centralQueue.poll()) != null) {
+            curr.release();
```

In wonder if we should guard against `IllegalReferenceException` here and so ensure we always poll and release everything even if an exception was thrown. In this case we should rethrow any exceptions once we drained the queue.